frame/support/src/traits/misc.rs

```diff
@@ -20,9 +20,187 @@
+///    message was emitted, then re-execute that block using `check-block` or `try-runtime`
+///    subcommands in substrate client.
+pub trait Defensive<T> {
+	/// Exactly the same as `unwrap_or`, but it does the defensive warnings explained in the trait
+	/// docs.
+	fn defensive_unwrap_or(self, other: T) -> T;
+	/// Exactly the same as `unwrap_or_else`, but it does the defensive warnings explained in the
+	/// trait docs.
+	fn defensive_unwrap_or_else<F: FnOnce() -> T>(self, f: F) -> T;
+	/// Exactly the same as `unwrap_or_default`, but it does the defensive warnings explained in the
+	/// trait docs.
+	fn defensive_unwrap_or_default(self) -> T
+	where
+		T: Default;
+}
+
+/// Same as [`Defensive`], but it fixes the second issue (logs not being expressive) at the cost of
+/// having a more verbose API.
+///
+/// Each defensive operation must contain a proof as well, as it should, in principle, be
+/// infallible. The API of this trait expects this proof to be given as a string, which is in turn
+/// printed in the emitting logs and panic stack trace. This can help with easier diagnosis.
+pub trait DefensiveWithProof<T> {
+	/// Exactly the same as `unwrap_or`, but it does the defensive warnings explained in the trait
+	/// docs.
+	fn defensive_unwrap_or_proof(self, other: T, proof: &'static str) -> T;
+	/// Exactly the same as `unwrap_or_else`, but it does the defensive warnings explained in the
+	/// trait docs.
+	fn defensive_unwrap_or_else_proof<F: FnOnce() -> T>(self, f: F, proof: &'static str) -> T;
+	/// Exactly the same as `unwrap_or_default`, but it does the defensive warnings explained in the
+	/// trait docs.
+	fn defensive_unwrap_or_default_proof(self, proof: &'static str) -> T
+	where
+		T: Default;
+}
+
+/// A variant of [`Defensive`] with the same rationale, for the arithmetic operations where in case
+/// an infallible operation fails, it saturates.
+pub trait DefensiveSaturating {
+	/// Add `self` and `other` defensively.
+	fn defensive_saturating_add(self, other: Self) -> Self;
+	/// Subtract `other` from `self` defensively.
+	fn defensive_saturating_sub(self, other: Self) -> Self;
+	/// Multiply `self` and `other` defensively.
+	fn defensive_saturating_mul(self, other: Self) -> Self;
+}
+
+// NOTE: A bit unfortunate, since T has to be bound by all the traits needed. Could make it
+// `DefensiveSaturating<T>` to mitigate, but anyways okay for a draft PR.
+impl<T: CheckedAdd + CheckedMul + CheckedSub + Bounded> DefensiveSaturating for T {
+	fn defensive_saturating_add(self, other: Self) -> Self {
+		self.checked_add(&other).defensive_unwrap_or_else(Bounded::max_value)
+	}
+	fn defensive_saturating_mul(self, other: Self) -> Self {
+		self.checked_sub(&other).defensive_unwrap_or_else(Bounded::max_value)
+	}
+	fn defensive_saturating_sub(self, other: Self) -> Self {
+		self.checked_mul(&other).defensive_unwrap_or_else(Bounded::min_value)
```

Comment:
```suggestion
		self.checked_sub(&other).defensive_unwrap_or_else(Bounded::min_value)
```

Reply:
holy smoke that could have been bad, thanks!